### PR TITLE
refactor(NODE-6057): implement CursorResponse for lazy document parsing

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

- Implement a `CursorResponse` subclass of MongoDBResponse
  - Implements logic to abstract List vs CursorResponse handling
- Make FindCursors use CursorResponse

##### Is there new documentation needed for these changes?

TBD. Is this something we want to apply to all cursors? Should we progressively migrate to this?

#### What is the motivation for this change?

An attempt at increasing performance. A sucessful decoupling of driver BSON and user BSON settings.

Only for FindCursor, BSON documents from the current batch are now parsed on each iteration. 

One clear improvement is that any cursor that is closed before exhaustion will take significantly less run time.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/client-side-encryption/auto_encrypter.ts [hunk 1]
```diff
@@ -473,8 +474,19 @@ export class AutoEncrypter {
   /**
    * Decrypt a command response
    */
-  async decrypt(response: Uint8Array | Document, options: CommandOptions = {}): Promise<Document> {
-    const buffer = Buffer.isBuffer(response) ? response : serialize(response, options);
+  async decrypt(
```
Comment (nbbeeken) [inline src/client-side-encryption/auto_encrypter.ts:477]: I can make FLE work or we can exempt it like we have to for explain

File: src/cursor/abstract_cursor.ts [hunk 2]
```diff
@@ -382,14 +389,7 @@ export abstract class AbstractCursor<
...
 
-    const doc = await next<TSchema>(this, { blocking: true, transform: false });
-
-    if (doc) {
-      this[kDocuments].unshift(doc);
-      return true;
-    }
-
-    return false;
+    return await next(this, { blocking: true, transform: false, hasNext: true });
```
Comment (nbbeeken) [inline src/cursor/abstract_cursor.ts:392]: If you look through my commits I originally supported unshift so that hasNext worked without changes but it was more logic than it was worth to hold on to the unshifted document. I think this is a good improvement regardless of CursorResponse, I wouldn't mind merging it separately

File: test/unit/assorted/collations.test.js [hunk 3]
```diff
@@ -55,7 +55,7 @@ describe('Collation', function () {
         request.reply(primary[0]);
       } else if (doc.aggregate) {
         commandResult = doc;
-        request.reply({ ok: 1, cursor: { id: 0, firstBatch: [], ns: 'collation_test' } });
+        request.reply({ ok: 1, cursor: { id: 0n, firstBatch: [], ns: 'collation_test' } });
```
Comment (nbbeeken) [inline test/unit/assorted/collations.test.js:58]: Should we use getFlexibleNumber? a real cursor id will ALWAYS be an int64

File: src/cmap/wire_protocol/responses.ts [hunk 4]
```diff
@@ -83,27 +95,98 @@ export class MongoDBResponse extends OnDemandDocument {
...
+  public batchSize = 0;
+
+  private batch: OnDemandDocument | null = null;
+  private values: Generator<OnDemandDocument, void, void> | null = null;
+  private iterated = 0;
+
+  constructor(b: Uint8Array, o?: number, a?: boolean) {
+    super(b, o, a);
+
+    if (this.isError) return;
```
Comment (nbbeeken) [inline src/cmap/wire_protocol/responses.ts:147]: I don't like this. I would like to potentially write a helper that figures out before calling the responseType constructor if it is an error and then never calls the passed in responseType since we will always convert an ok:0 to a MongoServerError.

File: src/cmap/wire_protocol/responses.ts [hunk 5]
```diff
@@ -83,27 +95,98 @@ export class MongoDBResponse extends OnDemandDocument {
...
+
+    const id = cursor.get('id', BSONType.long, true);
+    this.id = new Long(Number(id & 0xffff_ffffn), Number((id >> 32n) & 0xffff_ffffn));
+
+    const namespace = cursor.get('ns', BSONType.string) ?? '';
+    if (namespace) this.ns = ns(namespace);
+
+    if (cursor.has('firstBatch')) this.batch = cursor.get('firstBatch', BSONType.array, true);
+    else if (cursor.has('nextBatch')) this.batch = cursor.get('nextBatch', BSONType.array, true);
+    else throw new MongoUnexpectedServerResponseError('Cursor document did not contain a batch');
```
Comment (nbbeeken) [inline src/cmap/wire_protocol/responses.ts:159]: This and the other bson `get` calls with `required: true` have the chance of reaching users IFF the server replies with a malformed response. I am still comfortable with this because previously we would have failed somewhere in cursor code if `response.cursor... etc` was undefined with a TypeError, now we would pin point the problem at the layer of BSON missing something we expected to always be there.

File: src/cursor/abstract_cursor.ts [hunk 6]
```diff
@@ -144,7 +145,13 @@ export abstract class AbstractCursor<
   /** @internal */
   [kNamespace]: MongoDBNamespace;
   /** @internal */
-  [kDocuments]: List<TSchema>;
+  [kDocuments]: {
+    length: number;
+    shift(bsonOptions?: any): TSchema | null;
```
Comment (nbbeeken) [inline src/cursor/abstract_cursor.ts:150]: List.shift will get bson options passed into it that it will make no use of. We won't accidentally introduce new list API usage that isn't supported by CursorResponse

File: src/cmap/wire_protocol/responses.ts [hunk 7]
```diff
@@ -83,27 +95,98 @@ export class MongoDBResponse extends OnDemandDocument {
...
+  public override toObject(options?: BSONSerializeOptions): Record<string, any> {
     const exactBSONOptions = {
-      useBigInt64: options.useBigInt64,
-      promoteLongs: options.promoteLongs,
-      promoteValues: options.promoteValues,
-      promoteBuffers: options.promoteBuffers,
-      bsonRegExp: options.bsonRegExp,
-      raw: options.raw ?? false,
-      fieldsAsRaw: options.fieldsAsRaw ?? {},
+      useBigInt64: options?.useBigInt64,
```
Comment (nbbeeken) [inline src/cmap/wire_protocol/responses.ts:100]: I ran the de-opt explorer over this, and since `{}` empty object was often provided then none of these keys exist.

File: src/client-side-encryption/auto_encrypter.ts (hunk 1 above)
Comment (baileympearson) [reply src/client-side-encryption/auto_encrypter.ts:477]: I don't feel strongly.  how much effort do you think it would be to implement for FLE as well?

Comment (nbbeeken) [reply src/client-side-encryption/auto_encrypter.ts:477]: Sorry this wasn't clear, it is working, the changes here make it work but I am offering to defer impacting FLE since originally we _only_ wanted to change the find cursor, I don't mind or feel strongly either way.

File: src/cursor/abstract_cursor.ts (hunk 2 above)
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:392]: Here's the commit where I removed it: [`329ca55` (#4085)](https://github.com/mongodb/node-mongodb-native/pull/4085/commits/329ca55bcd58d5c8c798cf25936cf2c18fd71a30)

I have to buffer the unshifted documents to be returned by shift at a later point, in my view there is complexity & risk to maintaining this look-aside list that must be iterated first to ensure we don't repeat parsing effort and maintain the correct order, better to just not pull things out of a list.

Happy to change the name and also move this change out of this PR to keep it isolated. `checkNext`? `iterate: false`? `dequeue: false`?

File: src/client-side-encryption/state_machine.ts [hunk 8]
```diff
@@ -156,14 +157,15 @@ export class StateMachine {
...
+    context: MongoCryptContext,
+    responseType?: MongoDBResponseConstructor
   ): Promise<T> {
     const keyVaultNamespace = executor._keyVaultNamespace;
     const keyVaultClient = executor._keyVaultClient;
     const metaDataClient = executor._metaDataClient;
     const mongocryptdClient = executor._mongocryptdClient;
     const mongocryptdManager = executor._mongocryptdManager;
-    let result: T | null = null;
+    let result: any | null = null;
```
Comment (addaleax) [inline src/client-side-encryption/state_machine.ts:168]: I mean, TypeScript would be kind of right to complain here? The return type doesn't necessarily `extends Document` anymore

I think for "proper" typing you'd need to have overloads that account for the presence or absence of `responseType`:

```ts
    async execute<R extends typeof MongoDBResponseConstructor & {new(): R['prototype']}>(..., responseType: R): Promise<R['prototype']>;
    async execute<T extends Document>(...): Promise<T>;
    async execute<T extends Document, R extends typeof MongoDBResponseConstructor & {new(): R['prototype']}>(..., responseType?: R): Promise<T | R['prototype']> {
```

([TS playground](https://www.typescriptlang.org/play?ssl=14&ssc=158&pln=11&pc=12#code/JYOwLgpgTgZghgYwgAgCIHsEFcC2FzIDeAUMmcgNoDWEAngFzIDOYUoA5gLqMtsjsBuYgF9ixUJFiIUAFXQAHADIQAbhAA2GbHgIQAHpBAATJmky58YIqXIB9YEZ6sOQ0cQTq4TUwFl0-dFQAIQAlCCZ5fyYIAGEo1iwEMHQoIjcPL1MY9VMScmQvWhAEZH0IbEgAHhDSg3wTZDBaeQh0GGQ-AOCwiKjY+KhE5NSAMiIQCAB3AAoASkYQigByeSh0ZKaWpc5hAD5pqHDIkGiZZogF+eQABTWcYGjq5dX19fPt3aF8wuLa8qwqjJaoYGloLOB9ldbuh7o8ZJ8bGQfiUyhUIJUgfoQaYwTowAAaZA1LH1UybVrtTrsQKhI59OInBJJFLIMaECYzK6LFZrDbvHb7Q69E4QM4tAD8l0Y0Nh6KBAB8ic9eW8tpxdtZ8uRgO0DnSRWKILNkIcwFgoCBkByTfrTuc5l8tTazRaiDB0OhGEsAEZwKBLYSOsiidLxZCgFhwYooAC8VqmyGyTAd7jD7vQyDjEbAUaQADpUQCINMqTSesdogzeEMUrMhAgw77UlnGbmIAW9P9ICmG4zkL6AF6Z8Ot6MdrtyhTKNSacx4yFCIA), wanted to check if this actually works 🙂 )

File: src/cmap/wire_protocol/responses.ts [hunk 9]
```diff
@@ -83,27 +95,98 @@ export class MongoDBResponse extends OnDemandDocument {
...
-
     return { utf8: { writeErrors: false } };
   }
 }
+
+/** @internal */
+export class CursorResponse extends MongoDBResponse {
+  static emptyGetMore = new CursorResponse(
+    Buffer.from(
+      'NgAAABBvawABAAAAA2N1cnNvcgAhAAAAEmlkAAAAAAAAAAAABG5leHRCYXRjaAAFAAAAAAAA',
```
[the remark this answers is not in the document]
Comment (addaleax) [reply src/cmap/wire_protocol/responses.ts:127]: Yeah, if not this, then leaving a comment with instructions for how to create this string might be a good idea :)

File: src/cursor/abstract_cursor.ts (hunk 2 above)
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:392]: WIP https://github.com/mongodb/node-mongodb-native/pull/4093

File: src/cmap/wire_protocol/responses.ts [hunk 10]
```diff
@@ -83,27 +95,98 @@ export class MongoDBResponse extends OnDemandDocument {
...
+
+    if (cursor.has('firstBatch')) this.batch = cursor.get('firstBatch', BSONType.array, true);
+    else if (cursor.has('nextBatch')) this.batch = cursor.get('nextBatch', BSONType.array, true);
+    else throw new MongoUnexpectedServerResponseError('Cursor document did not contain a batch');
+
+    this.batchSize = this.batch.size();
+  }
+
+  get length() {
+    return Math.max(this.batchSize - this.iterated, 0);
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts:165]: batchSize is the size of this batch, not the requested amount

File: src/cmap/wire_protocol/responses.ts [hunk 11]
```diff
@@ -83,27 +95,98 @@ export class MongoDBResponse extends OnDemandDocument {
...
+
+  constructor(b: Uint8Array, o?: number, a?: boolean) {
+    super(b, o, a);
+
+    if (this.isError) return;
+
+    const cursor = this.get('cursor', BSONType.object, true);
+
+    const id = cursor.get('id', BSONType.long, true);
+    this.id = new Long(Number(id & 0xffff_ffffn), Number((id >> 32n) & 0xffff_ffffn));
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts:152]: Creating a Long from the BigInt, apply a 32bit mask to the bottom of the number, then shift down and apply the same mask.

File: src/cmap/wire_protocol/responses.ts [hunk 12]
```diff
@@ -83,27 +95,98 @@ export class MongoDBResponse extends OnDemandDocument {
...
+    this.batchSize = this.batch.size();
+  }
+
+  get length() {
+    return Math.max(this.batchSize - this.iterated, 0);
+  }
+
+  shift(options?: BSONSerializeOptions): any {
+    this.iterated += 1;
+    this.values ??= this.batch?.valuesAs(BSONType.object) ?? null;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts:170]: We cannot because of the isError check, I left a comment on that line, I'll experiment with how that can be removed but if I recall my last response PR work correctly I don't think it is easy unless we move writeConcernErrors out to the operation layer

File: src/cmap/wire_protocol/responses.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/cmap/wire_protocol/responses.ts:152]: `Long.fromBigInt`?

File: src/cmap/wire_protocol/responses.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts]: done

File: src/cmap/wire_protocol/responses.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts:183]: Added comment

File: src/cmap/wire_protocol/responses.ts (hunk 4 above)
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts]: Fixed

File: src/cmap/wire_protocol/responses.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts]: This is fixed without changing anything related to wc errors

File: src/cmap/wire_protocol/responses.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/cmap/wire_protocol/responses.ts:205]: Why not fix BSON's fromBigInt implementation and then use fromBigInt here?

[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts:205]: We can, I'll file a ticket but there are requirements around using bigint literals in BSON that make this a bit difficult to implement efficiently over there.

File: src/client-side-encryption/auto_encrypter.ts [hunk 13]
```diff
@@ -473,8 +474,19 @@ export class AutoEncrypter {
...
+    response: Uint8Array | Document | MongoDBResponse,
+    options: CommandOptions = {}
+  ): Promise<Document> {
+    const buffer = MongoDBResponse.is(response)
+      ? response.toBytes()
+      : Buffer.isBuffer(response)
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/client-side-encryption/auto_encrypter.ts]: I don't feel strongly, but BSON's types specify `Uint8Array`s but detect runtime and return `Buffer`s if running in Nodejs.  Since our FLE bindings only support Nodejs, using `Buffer` APIs is safe

File: src/cmap/wire_protocol/responses.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/cmap/wire_protocol/responses.ts:183]: Any reason you chose a comment over just serializing the document?

[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts:183]: Running the BSON serializer at import time seems undesirable to me. There _should_ be nothing wrong with this but it's opting into more than logic than I think is necessary for this task.

Ideally, I don't need to make this instance at all maybe there's something more clever than hard-coded bytes here. 🤔

[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/wire_protocol/responses.ts]: Avoiding making a CursorResponse altogether seems like a win. Our AbstractCursor lowered its "contract" requirements from List to just `length and .shift` (still has the push(Many) methods but it will not call them when `CursorResponse.is()` is true)

File: src/client-side-encryption/auto_encrypter.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/client-side-encryption/auto_encrypter.ts]: "Safe" within some degree of assumption about the environment you are running in. I can make BSON in Node.js return Uint8Arrays:

```js
Object.defineProperty(Buffer.prototype, '_isBuffer', {value: true}); 
Buffer.isBuffer(require('bson').serialize({a:1}));
```

File: test/integration/client-side-encryption/driver.test.ts [hunk 14]
```diff
@@ -297,4 +297,111 @@ describe('Client Side Encryption Functional', function () {
       });
     });
   });
+
+  describe(
```
Comment (baileympearson) [inline test/integration/client-side-encryption/driver.test.ts:301]: Is auto_encrypter.test.ts a better place for this test?

File: test/integration/client-side-encryption/driver.test.ts [hunk 15]
```diff
@@ -297,4 +297,111 @@ describe('Client Side Encryption Functional', function () {
...
+    'when @@mdb.decorateDecryptionResult is set on autoEncrypter',
+    { requires: { mongodb: '>=7.0' } },
```
Comment (baileympearson) [inline test/integration/client-side-encryption/driver.test.ts:303]: What about this test requires >= 7.0?

File: src/operations/find.ts [hunk 16]
```diff
@@ -77,13 +77,10 @@ export class FindOperation extends CommandOperation<Document> {
...
   filter: Document;
 
-  constructor(
-    collection: Collection | undefined,
-    ns: MongoDBNamespace,
-    filter: Document = {},
-    options: FindOptions = {}
-  ) {
-    super(collection, options);
+  public encryptionEnabled = false;
```
Comment (baileympearson) [inline src/operations/find.ts:80]: why is this necessary?  without this, when encryption is enabled:

- the find operation will provide a MonogoDBResponse to the server
- the server will provide it to the connection
- the crypto connection won't pass it along, so a raw response is returned

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/driver.test.ts (hunk 15 above)
Comment (nbbeeken) [reply test/integration/client-side-encryption/driver.test.ts]: https://spruce.mongodb.com/task/mongo_node_driver_next_rhel80_large_Node20_test_4.0_sharded_cluster_patch_ca2bfb0c536b665862a944c214e7cc51b49dbf1e_662fd327eb2a880007c01c39_24_04_29_17_04_51/logs?execution=0&sortBy=STATUS&sortDir=ASC

7.0 was too agressive, the $jsonSchema format I've used can't be used on 4.0 or below. I tried figuring out when this was added but it is not clear to me, do you happen to know? anyway lowered the filter requirement.

File: test/integration/client-side-encryption/driver.test.ts (hunk 14 above)
Comment (nbbeeken) [reply test/integration/client-side-encryption/driver.test.ts:301]: In the unit directory? I set this up to be integration because the unit coverage we already have did not catch the issue introduced, I want to know that `find` will return the right object

File: src/operations/find.ts (hunk 16 above)
Comment (nbbeeken) [reply src/operations/find.ts]: You're right, I think I did this in parallel with removing the pass through for crypto and didn't go back, good catch ty!

File: src/client-side-encryption/auto_encrypter.ts (hunk 1 above)
Comment (nbbeeken) [reply src/client-side-encryption/auto_encrypter.ts]: Yea I think we do, and I think we should get FLE on to the response train 🚂 but I would need to refactor the state machine and the decorate logic a bit more than is comfortable for this PR, so I think we're better off exempting crypto here and we can refactor in a follow up

**Review by baileympearson — APPROVED** [review]
